### DISCUSSION TITLE
how to pass parameters to range function in a query

### DISCUSSION THREAD

**Discussion by tonne1** [duckdb/duckdb#12275] [Q&A]
Gday, the following table function generates a timetable for a year:

```sql
create or replace function timetable (yr) as table
select range a_date from range(make_date(yr, 1, 1), make_date(yr + 1, 1, 1), INTERVAL 1 DAY);
``` 

This runs fine:

```sql
select * from timetable(2024);
``` 

But trying to use this within a query:

```sql
select b.a_date 
from (select 2024) a(yr)
cross join timetable(a.yr) b;
``` 

returns

SQL Error: java.sql.SQLException: Binder Error: 
Table function "range" does not support lateral join column parameters - cannot use column "a.yr" in this context

A workaround of course would be to setup range with static limits and then filter using a where part. There sure must be a better way. Thanks!

**Comment by soerenwolfers** [comment]
See https://github.com/duckdb/duckdb/issues/10927

**Comment by tonne1** [comment]
Thanks, @soerenwolfers! #10927 confirms that this is currently not supported. Well, if the table function is written slightly different, it does indeed support lateral join column parameters:

```sql
create or replace function timetabley(yr) as table
from unnest(range(make_date(yr, 1, 1), make_date(yr + 1, 1, 1), INTERVAL 1 DAY)) t(a_date);
```
And now, this works as expected:

```sql
select b.a_date 
from (select 2024) a(yr)
cross join timetable(a.yr) b;
```

As does this:

```sql
from timetabley(2024)
```
and this:

```sql
from timetabley(unnest([2024, 2025]))
```
But most importantly, it runs in a regular query context where other tables provide the input to the table function.
The main difference to the original version is that the result of the range function is wrapped with unnest. This workaround is not obvious to me. I can only guess that because unnest is one of a few table-in-out functions, the macro inherits that too.

There are numerous issues and discussions along that topic (#11338, #9995, #9998, #7846, #7522, #10927, #11393, #5485, #6513) where people have the same difficulties to see that duckdb does not support passing arguments in a straightforward way. As end user, I don't need to be concerned with internal representations. The table function receives a tuple as input and returns a dataset as output, why should it not be used in a simple standard query? This defeats aiming at "Friendly SQL".

Coming from other RDBMSes, if you define a function oder procedure, it's understood that you can use it in ANY context. For me personally, that's the top issue with duckdb, it would be much more useful if parameter handling was orthogonal. Apart from that, it's a great machine, thanks for that to everyone contributing!

**Reply by soerenwolfers** [reply]
Huh, I'm really surprised your workaround works. Especially since there were previous attempts to solve the table in out problem that were ultimately fruitless so far, and I don't understand now why they can't do your workaround under the hood. Anyway, good to know, thanks for sharing.
